cvnp/cvnp.cpp

```diff
@@ -9,7 +9,59 @@ namespace cvnp
     namespace detail
     {
         namespace py = pybind11;
-        
+
+        // Translated from cv2_numpy.cpp in OpenCV source code
+        class CvnpAllocator : public cv::MatAllocator {
+        public:
+            CvnpAllocator() { stdAllocator = cv::Mat::getStdAllocator(); }
+            ~CvnpAllocator() {}
+
+            cv::UMatData* allocate(pybind11::array& a) const {
+                //int dims, const int* sizes, int type, size_t* step) const {
+                cv::UMatData* u = new cv::UMatData(this);
+                u->data = u->origdata = (uchar*)a.mutable_data(0);
+                // auto *strides = a.strides();
+                // for( int i = 0; i < dims - 1; i++ )
+                //     step[i] = (size_t)_strides[i];
+                // step[dims-1] = CV_ELEM_SIZE(type);
+                // u->size = sizes[0]*step[0];
+                u->size = a.size();
+                u->userdata = a.inc_ref().ptr();
+                return u;
+            }
+
+            cv::UMatData* allocate(int dims0, const int* sizes, int type, void* data, size_t* step, cv::AccessFlag flags, cv::UMatUsageFlags usageFlags) const override
+            {
+                // TODO: I don't think this is ever called
```

Comment:
I think OpenCV's numpy code has an implementation of this because when they do a copy they set this as the allocator and call mat.copyTo... but we don't have an equivalent here.

Reply:
This is right. In my implementation, I was calling Mat.create(), which called this. In your case, you are directly calling your own version of "allocate(pybind11::array)".